lib/puppet/provider/group/windows_adsi.rb

```diff
@@ -67,7 +67,10 @@ def group
 
   def members
     @members ||= Puppet::Util::Windows::ADSI::Group.name_sid_hash(group.members)
-    @members.keys
+
+    # @members.keys returns an array of SIDs. We need to convert those SIDs into
+    # names so that `puppet resource` prints the right output.
+    members_to_s(@members.keys).split(',')
```

Comment:
This seems odd to me. The `user` provider's `groups` method doesn't translate to names: https://github.com/puppetlabs/puppet/blob/d82634c9427e3cc343b06b00677c7c3b72b1c74d/lib/puppet/provider/user/windows_adsi.rb#L20-L23

My mental model is that provider's getter and setter methods (for each property) should return the canonical value, as a manifest may contain duplicates, eg `'LocalSystem', 'S-1-5-18'`. And the insync method should compare canonical SIDs. And the `<property>_to_s` methods should convert SID to name.

Reply:
It is odd but this is what the `members` property used to do. The `groups` property also [returns names instead of SIDs](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type/user.rb#L346-L350). Unfortunately, the `members` property is implemented differently than the `groups` property (e.g. the `members` property does not override retrieve).

The `insync?` method does do the right thing for duplicates. Here's what that looks like: https://github.com/puppetlabs/puppet/blob/d82634c9427e3cc343b06b00677c7c3b72b1c74d/lib/puppet/provider/group/windows_adsi.rb#L27-L28

and `name_sid_hash`:
https://github.com/puppetlabs/puppet/blob/d82634c9427e3cc343b06b00677c7c3b72b1c74d/lib/puppet/util/windows/adsi.rb#L158-L168

(which takes care of duplicates in the manifest).